staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go

```diff
@@ -0,0 +1,322 @@
+	// Schema is a structural schema for this CELSchemaContext node. It is null if the current schema
+	// has no x-kubernetes-validations rules at or above this level or if the schema is not a valid structural schema.
+	Schema *structuralschema.Structural
+	// DeclType is a CEL declaration representation of Schema of this CELSchemaContext node. It is null if the current schema
+	// has no x-kubernetes-validations rules at or above this level or if the schema is not a valid structural schema.
+	DeclType *model.DeclType
+}
+
+// Converter converts from JSON schema to a structural schema and a CEL declType, or returns an error if the conversion
+// fails.
+type Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*CELTypeInfo, error)
+
+// DefaultConverter converts using standard conversion utilities.
+func DefaultConverter(schema *apiextensions.JSONSchemaProps, isRoot bool) (*CELTypeInfo, error) {
+	structural, err := structuralschema.NewStructural(schema)
+	if err != nil {
+		return nil, err
+	}
+	declType := model.SchemaDeclType(structural, isRoot)
+	if declType == nil {
+		return nil, fmt.Errorf("unable to convert structural schema to CEL declarations")
+	}
+	return &CELTypeInfo{structural, declType}, nil
+}
+
+// RootCelContext constructs CELSchemaContext for the given root schema.
+func RootCelContext(schema *apiextensions.JSONSchemaProps, converter Converter) CELSchemaContext {
+	rootCardinality := uint64(1)
+	r := CELSchemaContext{
+		jsonSchema:                schema,
+		withinValidationRuleScope: len(schema.XValidations) > 0,
+		MaxCardinality:            &rootCardinality,
+		TotalCost:                 &TotalCost{},
+		converter:                 converter,
+	}
+	return r
+}
+
+// ChildPropertyContext constructs and returns a CELSchemaContext for propertyName.
+func (c *CELSchemaContext) ChildPropertyContext(propSchema *apiextensions.JSONSchemaProps, propertyName string) CELSchemaContext {
+	return c.childContext(propSchema, propertyTypeInfoAccessor{propertyName: propertyName})
+}
+
+// ChildAdditionalPropertiesContext constructs and returns a CELSchemaContext for the properties of an object if this CELSchemaContext is an object.
+// schema must be non-nil and have a non-nil schema.AdditionalProperties.
+func (c *CELSchemaContext) ChildAdditionalPropertiesContext(propsSchema *apiextensions.JSONSchemaProps) CELSchemaContext {
+	return c.childContext(propsSchema, additionalItemsTypeInfoAccessor{})
+}
+
+// ChildItemsContext constructs and returns a CELSchemaContext for the items of an array if this CELSchemaContext is an array.
+func (c *CELSchemaContext) ChildItemsContext(itemsSchema *apiextensions.JSONSchemaProps) CELSchemaContext {
+	return c.childContext(itemsSchema, itemsTypeInfoAccessor{})
+}
+
+// ChildSchemaContext returns a CELSchemaContext for all child nodes that are not handled by ChildPropertiesContext,
+// ChildAdditionalPropertiesContext or ChildItemsContext such as not, allOf, oneOf and anyOf.
```

Comment:
I'm a little confused why we need this bit... I thought x-kubernetes-validations weren't allowed within not / allOf / oneOf / anyOf ... trying it, I got this:

```
* spec.validation.openAPIV3Schema.properties[a].allOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural
* spec.validation.openAPIV3Schema.properties[a].anyOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural
* spec.validation.openAPIV3Schema.properties[a].not.x-kubernetes-validations: Forbidden: must be empty to be structural
* spec.validation.openAPIV3Schema.properties[a].oneOf[0].x-kubernetes-validations: Forbidden: must be empty to be structural
```

... so there's no rules we'll need to attach cost computations to under these... can we just pass nil for celContext to ValidateCustomResourceDefinitionOpenAPISchema for these?

Reply:
We don't allow it.  I'll clean this up.